cmd/ocsp-updater/main.go

```diff
@@ -107,96 +243,162 @@ func (updater *OCSPUpdater) processResponse(tx *gorp.Transaction, serial string)
+	c := time.Tick(updater.missingSCTWindow)
+	for now := range c {
+		updater.rpcMu.RLock()
+
+		since := now.Add(-updater.oldestIssuedSCT)
+		serials, err := updater.getSerialsIssuedSince(since, updater.missingSCTBatch)
+		if err != nil && err != errNoNewCertificates {
+			updater.log.AuditErr(err)
+			updater.rpcMu.RUnlock()
+			continue
+		}
+
+		for _, serial := range serials {
+			// TODO(rolandshoemaker): For now this will do, in the future we should
+			// really have another method that allows us to specify which log/s should
+			// be submitted to so we don't double submit to logs we already have receipts
+			// for
+			if count, err := updater.getNumberOfReceipts(serial); err == nil && count != updater.numLogs {
+				var certDER []byte
+				updater.dbMap.SelectOne(
+					&certDER,
+					`SELECT der FROM certificates
+					 WHERE serial = :serial`,
+					map[string]interface{}{"serial": serial},
+				)
+				if err != nil {
+					updater.log.AuditErr(err)
+					continue
+				}
+
+				err = updater.pub.SubmitToCT(certDER)
+				if err != nil {
+					updater.log.AuditErr(err)
+					continue
+				}
+			} else if err != nil {
+				updater.log.AuditErr(err)
+				continue
 			}
 		}
 
-		updater.stats.TimingDuration("OCSP.Updates.BatchLatency", time.Since(outerStart), 1.0)
-		updater.stats.Inc("OCSP.Updates.BatchesProcessed", 1, 1.0)
+		updater.stats.TimingDuration("OCSP.TickDuration.MissingSCT", time.Since(now), 1.0)
+		updater.rpcMu.RUnlock()
 	}
-
-	return err
 }
 
-func main() {
-	app := cmd.NewAppShell("ocsp-updater", "Generates and updates OCSP responses")
+func (updater *OCSPUpdater) oldOCSPLoop() {
+	c := time.Tick(updater.oldOCSPWindow)
+	for now := range c {
+		updater.rpcMu.RLock()
+
+		certs, err := updater.findStaleOCSPResponses(now.Add(-updater.ocspMinTimeToExpiry))
+		if err != nil && err != errNoStaleResponses {
+			updater.stats.Inc("OCSP.Updates.Failed", 1, 1.0)
+			updater.log.AuditErr(err)
+			updater.rpcMu.RUnlock()
+			continue
+		}
 
-	app.App.Flags = append(app.App.Flags, cli.IntFlag{
-		Name:   "limit",
-		Value:  100,
-		EnvVar: "OCSP_LIMIT",
-		Usage:  "Count of responses to process per run",
-	})
+		tx, err := updater.dbMap.Begin()
+		if err != nil {
+			updater.stats.Inc("OCSP.Updates.Failed", 1, 1.0)
+			updater.log.AuditErr(err)
+			updater.rpcMu.RUnlock()
+			continue
+		}
+
+		failed := false
+		for _, cert := range certs {
+			err = updater.generateResponse(tx, cert)
+			if err != nil {
+				updater.stats.Inc("OCSP.Updates.Failed", 1, 1.0)
+				updater.log.AuditErr(err)
+				tx.Rollback()
+				failed = true
+				break
+			}
+			updater.stats.Inc("OCSP.Updates.Processed", 1, 1.0)
+		}
+
+		if !failed {
+			err = tx.Commit()
+			if err != nil {
+				updater.log.AuditErr(err)
+			}
+		}
 
-	app.Config = func(c *cli.Context, config cmd.Config) cmd.Config {
-		config.OCSPUpdater.ResponseLimit = c.GlobalInt("limit")
-		return config
+		updater.stats.TimingDuration("OCSP.TickDuration.OldOCSP", time.Since(now), 1.0)
```

Comment:
Okay, I've looked up Tick and Ticker.

This metric will not be all that we want. This will measure how long it took for us to perform the task from when the tick fired, but, since Tick can It adjusts the intervals or drops ticks to make up for slow receivers.", that metric will not be consistent with how badly we're performing.

We'll have to measure how long each call through the tick loop takes and emit it. As a convenience, it would be nice for this code to increment a simple counter when the time for a loop takes near, equal or more the tick duration.